File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int n = nums.size();
        vector<int> hashTable(n, 0);
        for(int i = 0; i < n; i++)
        {
            hashTable[nums[i]]++;
            if(hashTable[nums[i]] == 2)
                return nums[i];
        }
        return -1;
    }
};

File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        int n = nums.length;
        
        int slow = nums[0], fast = nums[0];
        
        do{
            slow = nums[slow];
            fast = nums[nums[fast]];
        }while(fast != slow);
        
        fast = nums[0];
        
        while(fast != slow)
        {
            slow = nums[slow];
            fast = nums[fast];
        }
        
        return slow;
    }
}